### PR TITLE
Prevent distortion of images with a 'sizes' attribute

### LINKED ISSUE #4606: Discontinue removing sizes attribute when converting img to amp-img
## Feature description

In #4548 the validator was updated to introduce new [`disable-inline-width` layout attribute](https://amp.dev/documentation/guides-and-tutorials/learn/amp-html-layout/#disable-inline-width). This will allow us to discontinue removing the `sizes` attribute on `amp-img` which the image sanitizer is currently doing:

https://github.com/ampproject/amp-wp/blob/014bbbba04bea102387f6faaa6b56ab8bf0e1e0a/includes/sanitizers/class-amp-img-sanitizer.php#L333-L334

This was confirmed in https://github.com/ampproject/amphtml/pull/27083#issuecomment-610767867.

The `sizes` attribute was removed to fix long-standing issues with images:

* https://github.com/ampproject/amp-wp/issues/1237
* https://github.com/ampproject/amphtml/issues/17053
* https://github.com/ampproject/amp-wp/issues/1086
* https://github.com/ampproject/amp-wp/issues/1305
* https://github.com/ampproject/amp-wp/pull/1939
* …

It was removed as part of a PR to “definitively” fix the issues: #2036. 

However, now that the `disable-inline-width` attribute is available to disable AMP's `sizes` behavior, we can add it automatically when converting `img` into `amp-img` (or `amp-anim`), and persist the `sizes` attribute as well.

This will involve re-checking the above issues for regressions!

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by pierlon**
## Summary

This PR re-adds the disabled `sizes` attribute to the `amp-img` component, and disables the inline width being added by adding the `disable-inline-width` attribute. This should prevent any distortion of the image when viewed with an AMP page.

I'm using the post content @westonruter posted in #2036 to test some of the possible ways an image can be used on a page. This verifies that there are no apparent regressions:

### Desktop

Non-AMP | AMP | AMP without fix
---|---|---
Screenshot_2020-04-23 Bison Image Test – WordPress Develop | Screenshot_2020-04-23 Bison Image Test – WordPress Develop | Screenshot_2020-04-23 Bison Image Test – WordPress Develop

### Mobile

Non-AMP | AMP | AMP without fix
---|---|---
Screenshot_2020-04-23 Bison Image Test – WordPress Develop(7) | Screenshot_2020-04-23 Bison Image Test – WordPress Develop(8) | Screenshot_2020-04-23 Bison Image Test – WordPress Develop(9)

Fixes #4606.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: includes/sanitizers/class-amp-img-sanitizer.php [hunk 1]
```diff
@@ -324,14 +325,15 @@ private function adjust_and_replace_node( $node ) {
...
 			} else {
-				$new_attributes['layout'] = 'intrinsic';
+				$new_attributes['layout'] = Layout::INTRINSIC;
 			}
 		}
 
-		// Remove sizes attribute since it causes headaches in AMP and because AMP will generate it for us. See <https://github.com/ampproject/amphtml/issues/21371>.
-		unset( $new_attributes['sizes'] );
+		if ( isset( $new_attributes['sizes'] ) ) {
+			$new_attributes['disable-inline-width'] = '';
```
Comment (schlessera) [inline includes/sanitizers/class-amp-img-sanitizer.php]: You should use `Attribute` here.
```suggestion
		if ( isset( $new_attributes[ Attribute::SIZES ] ) ) {
			$new_attributes[ Attribute::DISABLE_INLINE_WIDTH'] = '';
```
Note: You'll have to add the `DISABLE_INLINE_WIDTH` constant to the `Layout` class as well.

File: includes/sanitizers/class-amp-img-sanitizer.php [hunk 2]
```diff
@@ -353,7 +355,7 @@ private function adjust_and_replace_node( $node ) {
 		 */
 		if ( $img_node->hasAttribute( 'style' ) ) {
 			$layout = $img_node->getAttribute( 'layout' );
-			if ( in_array( $layout, [ 'fixed-height', 'responsive', 'fill', 'flex-item' ], true ) ) {
+			if ( in_array( $layout, [ Layout::FIXED_HEIGHT, Layout::RESPONSIVE, Layout::FILL, Layout::FLEX_ITEM ], true ) ) {
 				$required_display = 'block';
 			} elseif ( 'nodisplay' === $layout ) {
```
Comment (schlessera) [inline includes/sanitizers/class-amp-img-sanitizer.php]:
```suggestion
			} elseif ( Layout::NODISPLAY === $layout ) {
```

File: includes/sanitizers/class-amp-img-sanitizer.php [hunk 3]
```diff
@@ -324,14 +325,15 @@ private function adjust_and_replace_node( $node ) {
 				&& 'figure' === $node->parentNode->nodeName
 				&& preg_match( '/(^|\s)(alignwide|alignfull)(\s|$)/', $node->parentNode->getAttribute( 'class' ) )
 			) {
-				$new_attributes['layout'] = 'responsive';
+				$new_attributes['layout'] = Layout::RESPONSIVE;
```
Comment (schlessera) [inline includes/sanitizers/class-amp-img-sanitizer.php]:
```suggestion
				$new_attributes[ Attribute::LAYOUT ] = Layout::RESPONSIVE;
```

File: includes/sanitizers/class-amp-img-sanitizer.php [hunk 4]
```diff
@@ -324,14 +325,15 @@ private function adjust_and_replace_node( $node ) {
...
 			} else {
-				$new_attributes['layout'] = 'intrinsic';
+				$new_attributes['layout'] = Layout::INTRINSIC;
```
Comment (schlessera) [inline includes/sanitizers/class-amp-img-sanitizer.php]:
```suggestion
				$new_attributes[ Attribute::LAYOUT ] = Layout::INTRINSIC;
```

File: includes/sanitizers/class-amp-img-sanitizer.php [hunk 5]
```diff
@@ -132,20 +133,20 @@ public function sanitize() {
 			} elseif ( $node->hasAttribute( 'layout' ) ) {
```
Comment (schlessera) [inline includes/sanitizers/class-amp-img-sanitizer.php]:
```suggestion
			} elseif ( $node->hasAttribute( Attribute::LAYOUT ) ) {
```

File: includes/sanitizers/class-amp-img-sanitizer.php [hunk 6]
```diff
@@ -132,20 +133,20 @@ public function sanitize() {
...
 				$layout = $node->getAttribute( 'layout' );
```
Comment (schlessera) [inline includes/sanitizers/class-amp-img-sanitizer.php]:
```suggestion
				$layout = $node->getAttribute( Attribute::LAYOUT );
```

**Review by schlessera — DISMISSED** [review]
All-in with the constants... :)

File: includes/sanitizers/class-amp-img-sanitizer.php (hunk 1 above)
Comment (pierlon) [reply includes/sanitizers/class-amp-img-sanitizer.php]: Constants everywhere :smile:!

File: includes/sanitizers/class-amp-img-sanitizer.php [hunk 7]
```diff
@@ -312,26 +315,27 @@ private function adjust_and_replace_node( $node ) {
...
-				$new_attributes['layout'] = 'intrinsic';
+				$new_attributes[ Attribute::LAYOUT ] = Layout::INTRINSIC;
 			}
 		}
 
-		// Remove sizes attribute since it causes headaches in AMP and because AMP will generate it for us. See <https://github.com/ampproject/amphtml/issues/21371>.
-		unset( $new_attributes['sizes'] );
+		if ( isset( $new_attributes[ Attribute::SIZES ] ) ) {
+			$new_attributes[ Attribute::DISABLE_INLINE_WIDTH ] = '';
+		}
```
Comment (pierlon) [inline includes/sanitizers/class-amp-img-sanitizer.php:338]: Relevant change :arrow_up:.

**Comment by westonruter** [comment]
As expected, on a page with images, the changes in this PR result in the `i-amphtml-no-boilerplate` attribute being removed, with this comment added to the end of the `head`:

```
AMP optimization could not be completed due to the following:
 - CannotRemoveBoilerplate: Cannot remove boilerplate as either heights, media or sizes attribute is set: <amp-img noloading="" width="1600" height="1200" src="https://wordpressdev.lndo.site/content/uploads/2011/07/dsc09114.j…>
 - CannotRemoveBoilerplate: Cannot remove boilerplate as either heights, media or sizes attribute is set: <amp-img noloading="" width="1600" height="1200" src="https://wordpressdev.lndo.site/content/uploads/2011/07/dsc09114.j…>
```

This is expected as this regression will then be addressed as part of #4482.

**Comment by westonruter** [comment]
I'm testing this with the issue in #1939 regarding inline images. In Twenty Nineteen, adding content like so:

> image

The non-AMP version and the AMP version on `develop` look like this:

> image

But with this PR there is a regression:

> image

This is due to the `amp-img` having the `disable-inline-width` attribute which causes the element to get [this AMP shared CSS rule](https://github.com/ampproject/amphtml/blob/97099ce8bf4826a08d964e7a02da55a7641b1090/css/ampshared.css#L59-L65) applied:

```css
.i-amphtml-layout-responsive,
[layout=responsive][width][height]:not(.i-amphtml-layout-responsive),
[width][height][sizes]:not(.i-amphtml-layout-responsive)
{
  display: block;
  position: relative;
}
```

If I'm not mistaken, I think this CSS may actually be not updated for to account for the `disable-inline-width` attribute. I think that the last selector should perhaps be changed to not apply with the `disable-inline-width` attribute is present:

```
[width][height][sizes]:not([disable-inline-width]):not(.i-amphtml-layout-responsive)
``` 

I'm checking with the AMP team: https://github.com/ampproject/amphtml/pull/27083#issuecomment-618634128.

**Comment by westonruter** [comment]
Note this issue did not show up in testing Twenty Twenty because it (erroneously) has a `display:block` style added to all images.

**Comment by westonruter** [comment]
The `display:block` style also has the effect of centering the image on the attachment template (at least on Twenty Seventeen). See #1237.

**Comment by westonruter** [comment]
Opened a PR to fix the the issue in `ampshared.css`: https://github.com/ampproject/amphtml/pull/28020

**Review by westonruter — APPROVED** [review]
Upstream PR merged: https://github.com/ampproject/amphtml/pull/28020

**Comment by westonruter** [comment]
@schlessera anything else on this PR to receive your ✅?

**Comment by westonruter** [comment]
I guess not 😄 I'll merge and you can amend as needed in #4482.

**Pull request merged**